Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -394,4 +424,41 @@ class ASHorizontalScrollCellManager: NSObject, UICollectionViewDelegate, UIColle
         let pageWidth = CGRectGetWidth(scrollView.frame)
         pageChangedHandler?(scrollView.contentOffset.x / pageWidth)
     }
+
+    func collectionView(collectionView: UICollectionView, showContextMenuFor indexPath: NSIndexPath) {
+        //Show a context menu with options for the topsite
+        let contentItem = content[indexPath.row]
+
+        let alertController = UIAlertController(title: contentItem.siteURL.absoluteString, message: nil, preferredStyle: .ActionSheet)
+
+        let cancelAction = UIAlertAction(title: NSLocalizedString("Cancel", comment: "Label for Cancel button"), style: .Cancel, handler: nil)
+        alertController.addAction(cancelAction)
+
+        let deleteAction = UIAlertAction(title: NSLocalizedString("Remove", comment: "Label for Remove button"), style: .Destructive, handler: { (alert: UIAlertAction) -> Void in
+            self.collectionView(collectionView, deleteItemAtIndexPath: indexPath)
+        })
+        alertController.addAction(deleteAction)
+        // Because ASHorizontalScrollCellManager is not a UIViewController we cannot present a AlertController on it.
+        self.presentActionMenuHandler?(alertController)
+    }
+
+    func collectionView(collectionView: UICollectionView, deleteItemAtIndexPath indexPath: NSIndexPath) {
+        let contentItem = self.content[indexPath.row]
+        let preDeleteSize = content.count
+
+        // Delete the item from the DB
+        self.deleteItemHandler?(contentItem.siteURL)
+
+        // Delete the item from the current data source and animate the change
+        collectionView.performBatchUpdates({
+            self.content = self.content.filter {$0 != contentItem}
+            // There are alwaays some weird timing issues related to the underlying content changing before we can delete.
```

Comment:
Ugh collection view updating is the worst. Possibly related to the weird timing issue you were seeing but there is a chance of a race condition here.

The `deleteItemHandler` calls an async database method to update the domain table to hide from top sites on a background thread while the main thread performs the UIKit action of removing the cell. Indirectly, the underlying array of this data source is assigned from the result of a database query for top sites which could cause us to be in an awkward state.

The safer thing to do is to keep the database the source of truth for this collection view. If we were using Core Data for example, we would connect a `NSFetchedResultsController` here and have the model change send out a `didChange` notification to tell us to update the view. Likewise, instead of us initiating the deletion from the manager, it should be driven by the change in the underlying data model (in this case the SQL table). In this case, you'll want to chain the animation or just a basic reload table after removing the site, invalidating top sites, and resetting the data underneath the top sites manager.